src/android/CameraPreviewFragment.java

```diff
@@ -222,27 +230,69 @@ public void deviceHasUltraWideCamera(HasUltraWideCameraCallback hasUltraWideCame
         hasUltraWideCameraCallback.onResult(defaultCamera == true && ultraWideCamera == true);
     }
 
-    public static Size calculateResolution(Context context, int targetSize) {
-        Size calculatedSize;
-        if (getScreenOrientation(context) == Configuration.ORIENTATION_PORTRAIT) {
-            calculatedSize = new Size((int) ((float) targetSize / ratio), targetSize);
-        } else {
-            calculatedSize = new Size(targetSize, (int) ((float) targetSize / ratio));
+    public static Size calculateResolution(Context context, int desiredWidthPx, double aspectRatio) {
+        // Get all supported JPEG output sizes
+        Size[] supportedSizes = getSupportedResolutions(context, CameraSelector.LENS_FACING_BACK);
+
+        // Collect only those sizes matching the exact ratio
+        List<Size> matchingResolutions = new ArrayList<>();
+        for (Size size : supportedSizes) {
+            // Calculate the aspect ratio for the current resolution
+            double calculatedAspectRatio = (double) size.getHeight() / size.getWidth();
+
+            // Check if the calculated aspect ratio is close enough to the requested aspect ratio
+            if (Math.abs(calculatedAspectRatio - (aspectRatio)) < 0.01) {
+                matchingResolutions.add(size);
+            }
+        }
+
+        // If no exact matches, consider all supported sizes
+        List<Size> candidateResolutions = matchingResolutions.isEmpty()
+                ? Arrays.asList(supportedSizes)
+                : matchingResolutions;
```

Comment:
```suggestion
        if (matchingResolutions.isEmpty()) {
            matchingResolutions = Arrays.asList(supportedSizes);
        }
```
And using matchingResolutions instead of candidateResolution below

Maybe we can have a better naming for the variable.